gcp_variant_transforms/bq_to_vcf.py

```diff
@@ -56,19 +63,137 @@ def run(argv=None):
+  # type: (argparse.Namespace, pipeline_options.PipelineOptions, str) -> None
+  """Runs BigQuery to VCF shards pipelines.
+
+  It reads the variants from BigQuery table, groups a collection of variants
+  within a contiguous region of the genome (the size of the collection is
+  adjustable through flag `--number_of_bases_per_shard`), sorts them, and then
+  writes to one VCF shard.
+
+  TODO(allieychen): Eventually, it also generates the meta information file and
+  data header file.
+  """
   bq_source = bigquery.BigQuerySource(
       query=_BASE_QUERY_TEMPLATE.format(
           INPUT_TABLE='.'.join(bigquery_util.parse_table_reference(
               known_args.input_table))),
       validate=True,
       use_standard_sql=True)
 
-  options = pipeline_options.PipelineOptions(pipeline_args)
   with beam.Pipeline(options=options) as p:
     _ = (p | 'ReadFromBigQuery ' >> beam.io.Read(bq_source)
          | bigquery_to_variant.BigQueryToVariant()
          | densify_variants.DensifyVariants()
-         | vcfio.WriteToVcf(known_args.output_file))
+         | 'PairVariantWithKey' >>
+         beam.Map(_pair_variant_with_key, known_args.number_of_bases_per_shard)
+         | 'GroupVariantsByKey' >> beam.GroupByKey()
+         | beam.ParDo(_get_file_path_and_sorted_variants, bq_to_vcf_temp_folder)
+         | vcfio.WriteVcfDataLines())
+
+
+def _compose_temp_files(project, output_file, bq_to_vcf_temp_folder):
+  # type: (str, str, str) -> None
+  """Composes intermediate files to one VCF file.
+
+  It composes VCF data shards to one VCF data file and deletes the intermediate
+  VCF shards.
+  TODO(allieychen): Eventually, it further consolidates the meta information,
+  data header line, and the composed VCF data file into the `output_file`.
+  TODO(allieychen): Move the composing logic into a separate library.
+  """
+  bucket_name, blob_prefix = gcsio.parse_gcs_path(bq_to_vcf_temp_folder)
+  client = storage.Client(project)
+  bucket = client.get_bucket(bucket_name)
+  composed_vcf_data_file = _compose_vcf_shards_to_one(bucket, blob_prefix)
+  output_file_blob = _create_blob(client, output_file)
+  output_file_blob.rewrite(composed_vcf_data_file)
+  bucket.delete_blobs(bucket.list_blobs(prefix=blob_prefix))
+
+
+def _get_file_path_and_sorted_variants((file_name, variants), file_path_prefix):
+  # type: (Tuple[str, List], str) -> Iterable[Tuple[str, List]]
+  """Returns the file path and the sorted variants.
+
+  Args:
+    file_name: The file name that associated with the `variants`, which is used
+      to form the file path where the `variants` are written to.
+    variants: A collection of variants within a contiguous region of the genome.
+    file_path_prefix: The common file prefix for all VCF files generated in the
+      pipeline. The files written will begin with this prefix, followed by the
+      `file_name`.
+  """
+  file_path = '/'.join([file_path_prefix, file_name])
```

Comment:
Please use FileSystems.join here as well.

Reply:
I didn't use FileSystems here since it throws a 
`NameError: global name 'filesystems' is not defined [while running 'ParDo(CallableWrapperDoFn)']`
I have to explicitly import `filesystems` inside this method to avoid the above error. Please let me know if you still prefer use `FileSystems.join`

Reply:
ah i see...let's continue to use FileSystems to be consistent. 
You can import it within the method and disable pylint warnings `pylint:disable=wrong-import-order,wrong-import-position`